test/simulation_test.py

```diff
@@ -0,0 +1,141 @@
+
+        ecm = self.simulation.get_ecm()
+        self.assertEqual(len(ecm.get_entities()), 2)
+
+        agent_entity = ecm.get_entities_of_type("agent")[0]
+        self.assertEqual(agent_entity.get_components(Name)[0].get_state(), "Agent1")
+        self.assertEqual(agent_entity.get_components(Type)[0].get_state(), "agent")
+
+    def test_step(self):
+        self._add_agent("Agent1")
+
+        # Simulating one step
+        self.simulation.step()
+
+        # Check if the simulation state is updated
+        sim_state = self.simulation.get_sim_state()
+        self.assertAlmostEqual(sim_state.sim_time, 0.1)
+        self.assertEqual(len(sim_state.ecm_state), 1)
+        self.assertIn("entities", sim_state.ecm_state)
+
+        # Check if the agent's state is updated
+        ecm = self.simulation.get_ecm()
+        entities = ecm.get_entities_of_type(self._AGENT_TYPE)
+        self.assertEqual(len(entities), 1)
+        agent_entity = entities[0]
+        agent_pose = agent_entity.get_components(Pose)[0].get_state()
+        self.assertAlmostEqual(agent_pose[0], 0.2)  # Linear x position after 1 step
+
+    def test_step_for(self):
+        self._add_agent("Agent1")
+
+        # Simulating for 2 seconds
+        self.simulation.step_for(duration=2.0)
+
+        # Check if the simulation state is updated
+        sim_state = self.simulation.get_sim_state()
+        self.assertAlmostEqual(sim_state.sim_time, 2.0)
+
+        # Check if the agent's state is updated
+        ecm = self.simulation.get_ecm()
+        entities = ecm.get_entities_of_type(self._AGENT_TYPE)
+        self.assertEqual(len(entities), 1)
+        agent_entity = entities[0]
+        agent_pose = agent_entity.get_components(Pose)[0].get_state()
+        self.assertAlmostEqual(agent_pose[0], 4.0)  # Linear x position after 2 seconds
+
+    def test_get_sim_states(self):
+        self._add_agent("Agent1")
+
+        # Simulating for 2 seconds
+        self.simulation.step_for(duration=2.0)
+
+        # Check if all the states of the simulation are returned
+        sim_states = self.simulation.get_sim_states()
+        self.assertEqual(len(sim_states), 21)  # 20 steps + initial state
+
+    def _move_forward_controller(self, duration, sim_state, entity, ecm):
+        """Agent controller."""
+        pose = entity.get_components(Pose)[0]
+        velocity = entity.get_components(Velocity)[0]
+
+        # Move the agent forward in X direction using the current velocity.
+        pose._position[0] += duration * velocity.get_linear_vel()[0]
+
+    def _dumb_get_state(self):
+        return {"key": "value"}
+
+    def _add_agent(self, name):
+        initial_state = AgentInitialState(name=name, position=[0, 0, 0],
+                                          rotation=[0, 0, 0, 0], linear_vel=2., angular_vel=0.)
+        controller = self._move_forward_controller
+        get_state = self._dumb_get_state
+        self.simulation.add_agent(initial_state, controller, get_state)
```

Comment:
```suggestion
    def _add_agent(self, name):
        initial_state = AgentInitialState(name=name, position=[0, 0, 0],
                                          rotation=[0, 0, 0, 0], linear_vel=2., angular_vel=0.)
        controller = self._move_forward_controller
        get_state = lambda : {"key": "value"} 
        self.simulation.add_agent(initial_state, controller, get_state)
```

Reply:
To save some vertical space.

Reply:
You may also consider making `_move_forward_controller` a closure within this function.

Reply:
It was my first approach but flake8 complained about using direct `lambdas` and that's why I created the method.
I will add the method within _add_agent method though.

Reply:
Done.